### PR TITLE
Implement support for Arrays

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Repro of some weird behavior: https://runkit.com/nullvoxpopuli/validated-changeset-with-deep-arrays

image

I have a complex object with objects as array entries that I'd like to manage with a changeset. :D 

the example I have at work is:

```ts
{
  // big object
  listOfStuff: [
     {
       ... stuff
     },
     {
        ... some stuff
     } 
   ]
}
```

I suppose creating arrays when the property is undefined doesn't really make sense, because you can't know if the user does indeed want an array, or a if they want an object with numbered keys that would behave kinda like an array.

I propose that if a value is an array, we set at indicies, rather than properties on an object

### CODE REVIEW

**Comment by NullVoxPopuli** [comment]
Not sure how you feel about the approach I took here, but the gist is that merging arrays by index is weird -- we need to convert the arrays to objects with index-keys, making the objects "ArrayLike" so that we can merge two objects, rather than two arrays, which we already know how to do. I added [some utilities in a new file](https://github.com/validated-changeset/validated-changeset/pull/104/files#diff-bd1ef4eb51ab1a6a4cb2e9cc832fe5f9bd737e7f24d559a57f58271aeba8801a) to help out with this.

Some outstanding questions:
 - should `changeset.get('some.path.to.array')` always return an array after setting items on the array? [example test](https://github.com/validated-changeset/validated-changeset/pull/104/files#diff-ec61822d929089017a892b4b047229377c9241f16b43252ab851ae39131cdf40R798)
  - should the representation of the set of changes be changed to an array to always match the (sub)content type?

File: test/index.test.ts
Comment (snewcomer) [inline test/index.test.ts:795]: Could we have an assertion where `get/set` is out of range?

File: test/index.test.ts
Comment (snewcomer) [inline test/index.test.ts:802]: Could we also have some assertions for...

- `changeset.changes`
- `changeset.isValid`
- `changeset.save`

**Review by snewcomer — COMMENTED** [review]
Thanks for this PR!  I think I am generally in favor of this PR.  A few more assertions to battle test this would be helpful.

Note we also have some preliminary checks before merging you can run locally - `npm run test:ember-changeset && npm run test:ember-changeset-validations`

File: src/utils/set-deep.ts
Comment (NullVoxPopuli) [reply src/utils/set-deep.ts:69]: >  Are you indicating target[prop] might be an array?

yup, the object-array thing of today works with array's index access, which is cool. _thanks implicit type casting_! (for reals tho, this is one of the few times where I like this part of JS lol)

File: test/index.test.ts
Comment (NullVoxPopuli) [reply test/index.test.ts:795]: on the lower bound, out of range is only negative, I can add tests for that, but on the upper range, there is no out of bounds, because this is how you'd add stuff.
image

**Comment by NullVoxPopuli** [comment]
Also added:
 - library tests to CI
 - `yarn test:debug:named "part of test name here"` (or I suppose npm run, if that's what you're in to :upside_down_face: ) -- I can never remember all the args needed to debug with jest

File: test/index.test.ts
Comment (snewcomer) [reply test/index.test.ts:795]: Oh right.  But do you think it is worthwhile adding a test for this custom implementation when contact.emails is only len 2 but we do the following?

`dummyChangeset.set('contact.emails.10', 'the_fred@email.com');`

Comment (NullVoxPopuli) [reply test/index.test.ts:795]: yup! I'm working on adding more tests now (as soon as I figure out how to make `changeset.get('some array').toEqual(some array)`

Comment (NullVoxPopuli) [reply test/index.test.ts:795]: TIL 
image

I think this should probs be a hard error? even though behavior is kinda undefined in native arrays?

File: src/utils/object-tree-node.ts
Comment (NullVoxPopuli) [inline src/utils/object-tree-node.ts:15]: this feels fishy -- all of this is to make jest's expect work... which I feel like is kinda lying -- cause if you console.log(changeset.get('some array'), you get the ObjectNodeProxy, which.. I think is fine and it makes sense that you'd have to call unwrap on it to get the human-readable value out?

File: test/index.test.ts
Comment (NullVoxPopuli) [reply test/index.test.ts:872]: added

File: test/index.test.ts
Comment (NullVoxPopuli) [inline test/index.test.ts:849]: unwrap required here. 

Is `get` intended to automatically unwrap always?

File: test/index.test.ts
Comment (NullVoxPopuli) [inline test/index.test.ts:826]: without the weird jest hacks, this would require .unwrap() to be equallable with an array.
I think the main issue comes from the original object the proxy wraps being an object -- I tried to find a way to change that with Proxy, but it seems the ability to change how an object is iterated over or perceived is deliberately not part of the Proxy spec -- which kinda makes sense, cause it'd probs be super confusing

File: test/index.test.ts
Comment (snewcomer) [inline test/index.test.ts:918]: ☝️

File: src/utils/object-tree-node.ts
Comment (snewcomer) [inline src/utils/object-tree-node.ts:134]: This seems safe for ember-data.  B/c we aren't using the Ember provided `isArray`.

**Review by snewcomer — COMMENTED** [review]
This is looking pretty sound!

File: test/index.test.ts
Comment (NullVoxPopuli) [reply test/index.test.ts:864]: legit, thanks!

File: test/index.test.ts
Comment (NullVoxPopuli) [reply test/index.test.ts:826]: it would! _most excellent_

Comment (NullVoxPopuli) [reply test/index.test.ts:826]: and yeah, I don't know of any situations where in-app, I'd want to access a non-leaf

File: src/utils/merge-deep.ts
Comment (NullVoxPopuli) [reply src/utils/merge-deep.ts:155]: yup!

File: test/index.test.ts
Comment (NullVoxPopuli) [reply test/index.test.ts:918]: gasp! thanks!

File: test/index.test.ts
Comment (NullVoxPopuli) [inline test/index.test.ts:899]: idk if this is technically "deleting", but it made me wonder if changeset needs some post-processing callback to run before "save".

cause for some uses, you may want to collapse all the falsey values in an array, like, 
```ts
data.array = data.array.filter(Boolean);
```
but I could also see some use cases where you'd want the falsey values

**Comment by NullVoxPopuli** [comment]
also, should this scenario be supported?
```ts
      it('can add a new object at once, and edit it', () => {
        const changeset = Changeset(initialData);

        changeset.set('contact.emails.1', {
          funEmail: 'fun@email.com',
          primary: 'primary@email.com'
        });

        expect(changeset.get('contact.emails.1.funEmail')).toEqual('fun@email.com');
        expect(changeset.get('contact.emails.1.primary')).toEqual('primary@email.com');
        expect(changeset.get('contact.emails').unwrap()).toEqual([
          { primary: 'bob@email.com' },
          { primary: 'primary@email.com', funEmail: 'fun@email.com' }
        ]);
        expect(changeset.changes).toEqual([
          {
            key: 'contact.emails.1',
            value: { funEmail: 'fun@email.com', primary: 'primary@email.com' }
          }
        ]);

        changeset.set('contact.emails.1.primary', 'primary2@email.com');

        expect(changeset.get('contact.emails.1.primary')).toEqual('primary2@email.com');
      });
```

```
  ● Unit | Utility | changeset › arrays of objects within nested objects › #set › can add a new object at once, and edit it

    expect(received).toEqual(expected) // deep equality

    Expected: "primary2@email.com"
    Received: undefined

       997 |         changeset.set('contact.emails.1.primary', 'primary2@email.com');
       998 | 
    >  999 |         expect(changeset.get('contact.emails.1.primary')).toEqual('primary2@email.com');
           |                                                           ^
      1000 |       });
      1001 |     });
      1002 |   });
```

**Comment by NullVoxPopuli** [comment]
looks like yes, based on other tests (`#execute keeys prototype of set object` and `#execute it works with nested keys'`

**Comment by NullVoxPopuli** [comment]
@snewcomer We should be good to go :+1:

File: src/utils/set-deep.ts
Comment (NullVoxPopuli) [inline src/utils/set-deep.ts:86]: I don't know if the object case needs to also be based off the index, but for the ArrayLike situation, this made sense

File: test/utils/set-deep.test.ts
Comment (NullVoxPopuli) [inline test/utils/set-deep.test.ts:231]: I had to add this test, because my intuitive feelings about how arrays should work wasn't working.

idk if this is intended, or why it would be intended, but intuitively, I would have thoughts this would have been the case:

```ts
// ...
    expect(value).toEqual({
      top: {
        name: new Change('zoo'),
        foo: new Change({
          other: 'baz'
        })
      }
    });
```
because the `foo` is an entirely new object,

and how this correlates to arrays: 
instead of
```ts
it('sets when the array is initially empty', () => {
  const objA = {
    top: []
  };
  let value = setDeep(objA, 'top.0.name', new Change('zoo'));

  expect(value).toEqual({
    top: [{ name: new Change('zoo') }]
  });

  value = setDeep(value, 'top.0.foo.other', new Change('baz'));

  expect(value).toEqual({
    top: [
      {
        name: new Change('zoo'),
        foo: {
          other: new Change('baz')
        }
      }
    ]
  });
 });
```
maybe: 
```ts
it('sets when the array is initially empty', () => {
  const objA = {
    top: []
  };
  let value = setDeep(objA, 'top.0.name', new Change('zoo'));

  expect(value).toEqual({
    top: [new Change({ name: new Change('zoo') })]
  });

  value = setDeep(value, 'top.0.foo.other', new Change('baz'));

  // the big difference here -- the first entry remains an entire single Change
  expect(value).toEqual({
    top: [
      new Change({
        name: 'zoo',
        foo: {
          other: 'baz'
        }
      })
    ]
  });
});
```

File: test/utils/set-deep.test.ts
Comment (NullVoxPopuli) [inline test/utils/set-deep.test.ts:255]: I've started using `describe` blocks for groups of this stuff so I can debug scoped down sections of the test suite, so:
```bash
yarn test:debug:named "arrays at various nestings within an object"
```
allows me to debug just this subset of tests

File: test/index.test.ts
Comment (snewcomer) [reply test/index.test.ts:899]: I think in this case, the recommended way would be to `.set('contact.emails', [])`?  I'm happy leaving the nulls in there if that is what the `set` did!

**Review by snewcomer — APPROVED** [review]
This is a really really nice PR.  Thank you and let me know when you are comfortable for me to merge and release!

**Comment by NullVoxPopuli** [comment]
Yeah, I'm good with merging. I think at this point, I need to get our internal work merged (pending releases of this, ember changeset, and ember validated changeset).

Anything I run in to, I can open bugfix prs for

🎉

File: test/index.test.ts
Comment (NullVoxPopuli) [reply test/index.test.ts:818]: added!

File: test/index.test.ts
Comment (NullVoxPopuli) [reply test/index.test.ts:874]: added!

**Comment by NullVoxPopuli** [comment]
I found an inconsistency depending on the depth of a change:
```ts
changeset.get('contacts.emails') // => requires unwrap
changeset.get('emails') // does not have unwrap
```

I've found that there is a use case for wanting to access non-leaf data: iterating over the data to display navigation to different forms that would open a form to edit each array entry -- so... 
I don't think we should block this PR, as it's gotten kinda big, _but_, I do worry than forcing `changeset.get` to return a that node proxy for all complex objects _could_ be a breaking change

**Comment by NullVoxPopuli** [comment]
Turns out, I'm getting faster at this :D 

We should be good to go again :D 
No breaking change needed, since the added proxy wrap in `get` is specifically for arrays

**Comment by NullVoxPopuli** [comment]
I just keep finding issues. O.o

@snewcomer I added another test for what I'm running in to now. Idk if you want to take a stab at it. 

But I wonder if some refactoring might need to happen in order to make this easier? 
I _feel_ like this should be an easy addition due to how similar array and object access is -- but it hasn't been. :thinking: 

We also need a consistent way to get the "current value of the subtree at 'this' path", like.. I was thinking maybe a `preview` method that also takes a property path but _always_ returns the source data merged with the changes known for that path.

My recent issues have come from top level properties having inconsistent need for unwrap. Example:
 - have `emails: []`, can access via `changeset.get('emails');`
  - make change to an item within that array
  - `changeset.get('emails')` now needs `unwrap()` (only because I added an Array.isArray check in `get` to return a proxy) -- but without it, there is no merge of data and changes. It seems only the ObjectTreeNode can give a preview.

**Comment by NullVoxPopuli** [comment]
Just found out there is _something_ in this PR that fixes normal object behavior. So maybe it is best to merge it and iterate later? thoughts?

File: test/index.test.ts
Comment (snewcomer) [inline test/index.test.ts:1035]: Does this test pass if you add these assertions (similar to the test below)?

```
expect(changeset.get('emails.0.fun')).toEqual('fun0@email.com');
expect(changeset.get('emails.0.primary')).toEqual('primary0@email.com');
```

Comment (NullVoxPopuli) [reply test/index.test.ts:1035]: oh interesting. This is the diff with those:
image

Comment (NullVoxPopuli) [reply test/index.test.ts:1035]: I think this is a more general bug though, and why my work stuff _can't_ use the published validated changeset. 

_sometimes_ when accessing data via `get`, an object returns.
In _this_ case, the object is just totally incorrect. :thinking:

Comment (snewcomer) [reply test/index.test.ts:1037]: Ok so same as the one below (was just checking if the `value` was the culprit).

Overall, this is a really great PR and I think we are really close.  I think we can focus on `.get` or `ObjectTreeNode` may need to be modified.  I can certainly jump in and see where the last few improvements need to be made.

Comment (NullVoxPopuli) [reply test/index.test.ts:1037]: I'll take another stab, and see what happens :D

Comment (NullVoxPopuli) [reply test/index.test.ts:1037]: fixed it -- wasn't so bad after all :D

File: src/index.ts
Comment (snewcomer) [inline src/index.ts:42]: ✂️ 

Should make sure this is caught by linting in CI

File: src/index.ts
Comment (snewcomer) [inline src/index.ts:967]: What is the value of `result` here for the array case?  I thought one of the if expressions above would have been tripped.  It looks like there might be some minor refactors here.

File: src/index.ts
Comment (snewcomer) [inline src/index.ts:979]: For the array case, do we need to return an ObjectTreeNode?  Just a little background just in case there is some intersting information for this discussion...

The main reason for the `ObjectTreeNode` is to ensure we don't lose sibling keys.  So an object with both `changes` and unchanged content are able to be retrieved at any arbitrary level of the object. 

```
{
  changeValue: ...,
  original: ... 
}
```

Now for the array scenario, in one case we have `changes` we want to return.  In the other, we have `content` we want to return.  However, in both cases, I imagine that the type is an array and we don't have to deal with merging sibling keys.

**Review by snewcomer — COMMENTED** [review]
I'm really glad the tests are passing!  Just a few questions/comments before we can get this in!

File: src/index.ts
Comment (NullVoxPopuli) [reply src/index.ts:42]: I can PR tomorrow that adds / updates linting in the CI config tomorrow?

File: src/index.ts
Comment (NullVoxPopuli) [reply src/index.ts:967]: `result` is `undefined` here. It happens mostly when working with new array entries or previously deleted entries, iirc

File: src/index.ts
Comment (NullVoxPopuli) [reply src/index.ts:979]: > For the array case, do we need to return an ObjectTreeNode? 

yeah, cause atm, ObjectTreeNode is the only thing that knows that changes are on arrays and can convert between arrays and array objects.

_however_ this changes the behavior of `changeset.get('prop.someArray')` where it now needs `unwrap`, where it previously did not -- but this is more consistent with the deep object changes.

> . However, in both cases, I imagine that the type is an array and we don't have to deal with merging sibling keys.

🤷 
The overloaded behavior of this method is dizzying to me :(

**Pull request merged**
